bdb/bdb_verify.c

```diff
@@ -273,245 +213,582 @@ static void printhex(SBUF2 *sb, int (*lua_callback)(void *, const char *),
-    int (*vtag_callback)(void *parm, void *dta, int *dtasz, uint8_t ver),
-    int (*add_blob_buffer_callback)(void *parm, void *dta, int dtasz,
-                                    int blobno),
-    void (*free_blob_buffer_callback)(void *parm),
-    unsigned long long (*verify_indexes_callback)(void *parm, void *dta,
-                                                  void *blob_parm),
-    void *callback_parm, 
-    int (*lua_callback)(void *, const char *), void *lua_params, 
-    void *callback_blob_buf, int progress_report_seconds,
-    int attempt_fix, unsigned int lid)
+static int bdb_verify_data_stripe(verify_common_t *par, int dtastripe, unsigned int lid)
 {
     DBC *cdata = NULL;
     DBC *ckey = NULL;
     DB *db;
-    DBC *cblob;
+    DBC *cblob = NULL;
     unsigned char databuf[17 * 1024];
     unsigned char keybuf[18 * 1024];
     unsigned char expected_keybuf[18 * 1024];
-    unsigned char verify_keybuf[18 * 1024];
-    int dtastripe;
-    DBT dbt_data = {0};
-    DBT dbt_key = {0};
-    DBT dbt_blob_key = {0}, dbt_blob_data = {0};
-    DBT dbt_dta_check_key = {0}, dbt_dta_check_data = {0};
-    int rc;
-    int ix;
-    int keylen;
-    unsigned long long has_keys;
-    unsigned long long genid, verify_genid;
+    int rc = 0;
     int blobsizes[16];
     int bloboffs[16];
     int nblobs = 0;
-    int blobno;
-    char dumbuf;
-    uint8_t ver;
-    int now, last;
-    int64_t nrecs = 0;
-    int nrecs_progress = 0;
-    unsigned long long genid_flipped;
-    int ret = 0;
+    int last = comdb2_time_epochms();
+    blob_buffer_t blob_buf[MAXBLOBS] = {{0}};
 
-    bzero(&dbt_data, sizeof(DBT));
-    bzero(&dbt_key, sizeof(DBT));
+    bdb_state_type *bdb_state = par->bdb_state;
 
+    DBT dbt_data = {0};
     dbt_data.flags = DB_DBT_USERMEM;
     dbt_data.ulen = sizeof(databuf);
     dbt_data.data = databuf;
+
+    DBT dbt_key = {0};
     dbt_key.flags = DB_DBT_USERMEM;
     dbt_key.ulen = sizeof(keybuf);
     dbt_key.data = keybuf;
 
-    bzero(&dbt_blob_key, sizeof(DBT));
-    bzero(&dbt_blob_data, sizeof(DBT));
+    db = bdb_state->dbp_data[0][dtastripe];
+    rc = db->paired_cursor_from_lid(db, lid, &cdata, 0);
+    if (rc) {
+        logmsg(LOGMSG_ERROR, "dtastripe %d cursor rc %d\n", dtastripe, rc);
+        return rc;
+    }
+    uint8_t ver;
+    rc = bdb_cget_unpack(bdb_state, cdata, &dbt_key, &dbt_data, &ver,
+                         DB_FIRST);
+    int atstart = comdb2_time_epochms();
+    int now = atstart;
+    logmsg(LOGMSG_DEBUG, "%p:%s Entering stripe=%d\n", (void *)pthread_self(), __func__, dtastripe);
+
+    while (rc == 0 && !par->client_dropped_connection) {
+        ATOMIC_ADD(par->records_processed, 1);
+        par->nrecs_progress++;
+
+        now = comdb2_time_epochms();
+        print_verify_progress(par, now);
+
+        /* check if comdb2sc is killed every 1000ms */
+        if (par->verify_mode == VERIFY_DEFAULT && (now - last) > 1000) {
```

Comment:
We still want to check this in parallel mode as well.

Reply:
Ok, will move into print_verify_progress. Somehow I did not think it was needed in parallel mode -- not sure why.